pkg/resourcegroup/controllers/resourcegroup/resourcegroup_controller.go

```diff
@@ -350,19 +350,31 @@ func (r *reconciler) computeStatus(
 // to the legacy status field. If conversion is not needed, the original status field is returned
 // instead.
 func ActuationStatusToLegacy(s v1alpha1.ResourceStatus) v1alpha1.Status {
-	if s.Status == v1alpha1.NotFound {
-		return v1alpha1.NotFound
+	// When actuation is unknown, return the latest computed status.
+	// This should only ever happen when upgrading from an old version.
+	// It means the applier was last run with code that predates the addition
+	// of the strategy, actuation, and reconcile fields.
+	if s.Actuation == "" {
+		return s.Status
 	}
 
-	if s.Actuation != "" &&
-		s.Actuation != v1alpha1.ActuationSucceeded {
-		return v1alpha1.Unknown
+	// When actuation has succeeded, we know the object spec is up to date with
+	// the latest source, so the latest computed status should be correct.
+	if s.Actuation == v1alpha1.ActuationSucceeded {
+		return s.Status
 	}
 
-	if s.Actuation == v1alpha1.ActuationSucceeded && s.Reconcile == v1alpha1.ReconcileSucceeded {
-		return v1alpha1.Current
+	// When the latest computed status is Terminating or NotFound, keep it,
+	// even if actuation is pending, skipped, or failed.
+	if s.Status == v1alpha1.NotFound || s.Status == v1alpha1.Terminating {
+		return s.Status
 	}
-	return s.Status
+
+	// When actuation is not successful (pending, skipped, failed), that means
+	// the object spec has not been updated to the desired spec from source yet,
+	// so the computed status can be ignored and the real status is Unknown,
+	// unless the latest computed status is NotFound or Terminating.
+	return v1alpha1.Unknown
```

Comment:
So given actuation was not successful:
- `InProgress`, `Current`, or `Failed` is replaced with `Unknown`
- `NotFound` and `Terminating` are retained

What is special about `NotFound`/`Terminating` that makes it honored? And why are each of the other statuses replaced with `Unknown`?

Reply:
The status field here is sorta overloaded. It's not just the current kstatus (spec -> status). It's also trying to express whether the desired state from source has been synced (source -> spec -> status).

So when kstatus computes to InProgress, Current, and Failed, those all depend on the spec being the desired state. But if the desired state is actually in the source, and it hasn't been actuated yet, then the ResourceGroup controller doesn't know what the desired state is and neither does Kubernetes. So those all need to map to Unknown.

But when kstatus computes to NotFound or Terminating, those are independent of the desired state and are still correct and valuable to expose to the user.

Reply:
I updated the PR description to clarify the fix.

Reply:
Do we keep `NotFound` and `Terminating` because we don't care whether it exists in source (desired state) or not anymore?

Reply:
Right. 

After a sync attempt succeeds, the applier will remove pruned objects from the inventory, so this edge case doesn't apply. 

This case only applies before syncing has started (before inv-add-task) and after syncing has started (after inv-add-task) but before its succeeded (before inv-set-task).

If the actuation status is Pending, Skipped, or Failed, then Kubernetes does not have the latest object spec yet, so the kstatus is mostly incorrect, relative to the desired state in the source. But NotFound and Terminating are always correct, because they're not dependent on Kubernetes having the latest spec from source.

Reply:
Put another way:
- If the object status is NotFound or Terminating, we should show NotFound or Terminating, whether or not the object exists in the source or has changed since it was last applied.
- If the object status is InProgress, Current, or Failed AND we know that the last attempt to apply or prune it Failed or was Skipped or is Pending, then the latest spec has not been persisted to Kubernetes yet, so the RGC doesn't know if the current spec on the cluster matches the desired spec in source or not, so we set the status to Unknown.